applications/serial_lte_modem/src/lwm2m_carrier/slm_at_carrier.c

```diff
@@ -0,0 +1,786 @@
+	}
+
+	ret = util_string_to_float_get(&at_param_list, 4, &speed_v);
+	if (ret) {
+		return ret;
+	}
+
+	ret = util_string_to_float_get(&at_param_list, 5, &uncertainty_h);
+	if (ret) {
+		return ret;
+	}
+
+	ret = util_string_to_float_get(&at_param_list, 6, &uncertainty_v);
+	if (ret) {
+		return ret;
+	}
+
+	return lwm2m_carrier_velocity_set(heading, speed_h, speed_v, uncertainty_h, uncertainty_v);
+}
+
+/* AT#XCARRIER="portfolio","create|read|write",<instance_id>[,<identity_type>[,<identity>]] */
+static int do_carrier_portfolio(void)
+{
+	int ret, param_count;
+	uint16_t instance_id, identity_type;
+	char operation[7], buffer[64];
+	size_t size = sizeof(operation);
+	uint16_t buf_len = sizeof(buffer);
+
+	ret = util_string_get(&at_param_list, 2, operation, &size);
+	if (ret) {
+		return ret;
+	}
+
+	ret = at_params_unsigned_short_get(&at_param_list, 3, &instance_id);
+	if (ret) {
+		return ret;
+	}
+
+	param_count = at_params_valid_count_get(&at_param_list);
+	if (param_count > 4) {
+		ret = at_params_unsigned_short_get(&at_param_list, 4, &identity_type);
+		if (ret) {
+			return ret;
+		}
+	}
+
+	if (slm_util_cmd_casecmp(operation, "READ")) {
+		ret = lwm2m_carrier_identity_read(instance_id, identity_type, buffer, &buf_len);
+		if (ret) {
+			return ret;
+		}
+
+		sprintf(rsp_buf, "\r\n#XCARRIER: %s\r\n", buffer);
+		rsp_send(rsp_buf, strlen(rsp_buf));
+
+		return 0;
+	} else if (slm_util_cmd_casecmp(operation, "WRITE")) {
+		size = sizeof(buffer);
+
+		ret = util_string_get(&at_param_list, 5, buffer, &size);
+		if (ret) {
+			return ret;
+		}
+
+		return lwm2m_carrier_identity_write(instance_id, identity_type, buffer);
+	} else if (slm_util_cmd_casecmp(operation, "CREATE")) {
+		return lwm2m_carrier_portfolio_instance_create(instance_id);
+	}
+
+	LOG_DBG("AT#XCARRIER=\"portfolio\" failed: invalid operation");
+
+	return -EINVAL;
+}
+
+/* AT#XCARRIER="reboot" */
+static int do_carrier_request_reboot(void)
+{
+	return lwm2m_carrier_request(LWM2M_CARRIER_REQUEST_REBOOT);
+}
+
+/* AT#XCARRIER="link_down" */
+static int do_carrier_request_link_down(void)
+{
+	return lwm2m_carrier_request(LWM2M_CARRIER_REQUEST_LINK_DOWN);
+}
+
+/* AT#XCARRIER="link_up" */
+static int do_carrier_request_link_up(void)
+{
+	return lwm2m_carrier_request(LWM2M_CARRIER_REQUEST_LINK_UP);
+}
+
+/**@brief API to handle Carrier AT command
+ */
+int handle_at_carrier(enum at_cmd_type cmd_type)
+{
+	int ret;
+	char op_str[SLM_CARRIER_OP_STR_MAX];
+	int size = sizeof(op_str);
+
+	if (cmd_type != AT_CMD_TYPE_SET_COMMAND) {
+		return -EINVAL;
+	}
+	ret = util_string_get(&at_param_list, 1, op_str, &size);
+	if (ret) {
+		return ret;
+	}
+	ret = -EINVAL;
+	for (int i = 0; i < CARRIER_OP_MAX; i++) {
+		if (slm_util_casecmp(op_str, carrier_op_list[i].op_str)) {
+			ret = carrier_op_list[i].handler();
+			break;
+		}
+	}
+
+	return ret;
+}
+
+int slm_at_carrier_init(void)
+{
+	carrier_data_mode_handler = NULL;
+
+	return 0;
+}
+
+int slm_at_carrier_uninit(void)
+{
+	return 0;
```

Comment:
The _init and _uninit are both dummy. SLM #XSLEEP/#XRESET would call _uinit then _init (upon wakeup). Both would cut LTE connection first. There might be a need to handle stop/re-start of carrier_lwm2m lib as well.